### PR TITLE
Optimize drawing horizontal/vertical lines

### PR DESCRIPTION
**Pull request by a-wi**
This is a working PR to benchmark new function to draw horizontal/vertical lines in a new (optimized) way and to check whether it helps in fixing [issue 18517](http://trac.wxwidgets.org/ticket/18517).

### CODE REVIEW

File: src/msw/utilsgui.cpp [hunk 1]
```diff
@@ -209,6 +209,32 @@ void wxDrawLine(HDC hdc, int x1, int y1, int x2, int y2)
...
+    }
+    else
+    {
+        SetRect(&r, x1 - w1, y1, x2 + w2, y2);
+    }
+
+    COLORREF bgColorOrig = ::GetBkColor(hdc);
+    ::SetBkColor(hdc, color);
+
+    ::ExtTextOut(hdc, 0, 0, ETO_OPAQUE, &r, L"", 0, NULL);
```
Comment (PBfordev) [inline src/msw/utilsgui.cpp]: Just a nitpick:

If anyone uses non-Unicode build of wxWidgets (`ExtTextOut` = `ExtTextOutA` expecting `LPCSTR`), which AFAIK is frown upon but still allowed, wouldn't  using unconditionally a wide-string literal here break the build?

While not pretty, perhaps using `::ExtTextOutW()`  would be better...

Comment (vadz) [reply src/msw/utilsgui.cpp]: Yes, we need to either use `::ExtTextOutW()` or `wxT("")` here.

File: src/msw/dc.cpp [hunk 2]
```diff
@@ -817,7 +817,30 @@ void wxMSWDCImpl::DoCrossHair(wxCoord x, wxCoord y)
 
 void wxMSWDCImpl::DoDrawLine(wxCoord x1, wxCoord y1, wxCoord x2, wxCoord y2)
 {
-    wxDrawLine(GetHdc(), XLOG2DEV(x1), YLOG2DEV(y1), XLOG2DEV(x2), YLOG2DEV(y2));
+#if 1
```
Comment (vadz) [inline src/msw/dc.cpp]: I'd get rid of this, the change would be simple enough to revert later anyhow.

File: src/msw/dc.cpp [hunk 3]
```diff
@@ -817,7 +817,30 @@ void wxMSWDCImpl::DoCrossHair(wxCoord x, wxCoord y)
...
+    // If we are going to draw a vertical or horizontal line with pen with solid colour and flat cap
+    if ( ( x1 == x2 || y1 == y2 ) &&
+         ( !m_pen.IsOk() || ( m_pen.GetColour().IsSolid() && ( m_pen.GetWidth() == 1 || m_pen.GetCap() == wxCAP_BUTT ) ) ) )
```
Comment (vadz) [inline src/msw/dc.cpp]: I think this should be
```suggestion
         ( !m_pen.IsOk() || ( m_pen.GetStyle() == wxPENSTYLE_SOLID && ( m_pen.GetWidth() == 1 || m_pen.GetCap() == wxCAP_BUTT ) ) ) )
```

I'd also like to wrap this line to fit in 80 columns.

File: src/msw/utilsgui.cpp [hunk 4]
```diff
@@ -209,6 +209,32 @@ void wxDrawLine(HDC hdc, int x1, int y1, int x2, int y2)
...
+    wxASSERT(x1 == x2 || y1 == y2);
+
+    int w1 = width / 2;
+    int w2 = width - w1;
+    RECT r;
+    if ( y1 == y2 )
+    {
+        if ( x1 == x2 )
+            return;
+        SetRect(&r, x1, y1 - w1, x2, y1 + w2);
```
Comment (vadz) [inline src/msw/utilsgui.cpp]:
```suggestion
        ::SetRect(&r, x1, y1 - w1, x2, y1 + w2);
```
Just to make it clear that it's a Win32 API function.

File: src/msw/utilsgui.cpp [hunk 5]
```diff
@@ -209,6 +209,32 @@ void wxDrawLine(HDC hdc, int x1, int y1, int x2, int y2)
...
+    }
+    else
+    {
+        SetRect(&r, x1 - w1, y1, x2 + w2, y2);
```
Comment (vadz) [inline src/msw/utilsgui.cpp]:
```suggestion
        ::SetRect(&r, x1 - w1, y1, x2 + w2, y2);
```

**Review by vadz — COMMENTED** [review]
Thanks, looks good to me except for the pen style check.

The new function also could be used in `wxMSWDCImpl::DoCrossHair()` and, a bit more importantly, `DrawRect()` and `DrawButtonFrame()` in `src/msw/anybutton.cpp`

**Pull request merged**